### PR TITLE
Vehicle keep driving

### PR DESCRIPTION
**Pull request by frackledust**
This PR contains following changes:
1. Vehicle now stops its movement at the end of the segment, instead of at the beginning of the next one, therefore it can in their next turn recalculate the alternatives and choose the best route to continue on.
2. A bug in fcds was fixed, so the distance in fcds is the same as the one calculated in advance_vehicle
3. Calculation of vehicle density for LoS was changed, so it's only calculated on the part of the segment in front of the vehicle. Parameter "count-vehicles-tolerance" was added for the tolerance (in seconds ahead of the vehicle's time offset) that the cars should be counted in.
4. Queues were added. Cars at the end of the segment are added to a queue and can only leave the segment if all cars in front of them already left the segment or became inactive. Queues are saved in case Simulation needs to be run from continue-from pickle file.
5. In memory Map is now updated based on current speeds on segments - these are calculated as the avarage speed of the cars on the segment. Parameter "map-update-freq-steps" defines how often map should be updated.
Updating for HDF5 needs to be added.
6. Parameters can be now loaded from config.json file. If this file is available in the working directory it would overwrite parameters used in console.
7. Fastest path now uses current-travel-time instead of speed as weight.
8. Alternatives are only calculated for the cars at the end of the segment to make simulation faster.

### CODE REVIEW

File: ruth/simulator/route.py [hunk 1]
```diff
@@ -26,139 +24,224 @@ def move_on_segment(
...
+        if vehicle not in queue:
+            current_vehicle_list.remove(vehicle)
+            new_fcds = advance_vehicle(vehicle, departure_time, gv_db)
+            fcds.extend(new_fcds)
+        elif vehicle == queue[0]:
+            # vehicle is the first one in the queue
+            current_vehicle_list.remove(vehicle)
+            new_fcds = advance_vehicle(vehicle, departure_time, gv_db)
+            fcds.extend(new_fcds)
+            was_moved = len(queue) == 0 or (vehicle != queue[0])
```
[the remark this answers is not in the document]
Comment (Michailidu) [reply ruth/simulator/route.py:232]: Queues might have been changed in advance_vehicle function above in the same branch, so I think we need to check if the vehicle had left the queue.

[the remark this answers is not in the document]
Comment (Kobzol) [reply ruth/simulator/route.py:232]: In this commit the function worked with a global variable, so it wasn't at all obvious that `queue` could change :) Luckily, that has changed in the meantime, so now it's a bit more visible at least, that the queue could have changed.

File: ruth/data/map.py [hunk 2]
```diff
@@ -65,11 +88,48 @@ def name(self):
...
+        nx.set_edge_attributes(self.network, values=travel_times, name="current_travel_time")
+
+    def update_current_speeds(self, segment_ids, speeds):
+        new_speeds = {}
+        new_travel_times = {}
+        lengths = nx.get_edge_attributes(self.simple_network, name='length')
+        for (node_from, node_to), speed in zip(segment_ids, speeds):
+            new_speeds[(node_from, node_to)] = speed
+            new_travel_times[(node_from, node_to)] = float(lengths[(node_from, node_to)]) / float(speed)
+        nx.set_edge_attributes(self.simple_network, values=new_speeds, name='current_speed')
```
[the remark this answers is not in the document]
Comment (frackledust) [reply ruth/data/map.py:95]: The current speed will be useful for the cpp/distributed alternatives.

[the remark this answers is not in the document]
Comment (Kobzol) [reply ruth/data/map.py:94]: Hmm. Does that mean that the cpp version computes its own edge labels by dividing the segment length by `current_speed`? That sounds a bit scary, as this logic might not be identical in the simulator and in alternatives.

If distributed alternatives are used, then we don't need the travel times, right? Because we don't even compute any shortest paths locally. Maybe we should somehow detect this, because it's needless work to compute it if we don't use it. At the very least, it would be nice to measure how long does it take to run this function :) If it's quick, then we don't have to worry about it.

File: ruth/data/map.py [hunk 3]
```diff
@@ -65,11 +76,46 @@ def name(self):
...
     def set_data_dir(self, path):
         if self.data_dir is not None:
             cl.warn(f"The data dir has changed from '{self.data_dir}' to '{path}.")
         self.data_dir = path
 
+    def osm_route_to_py_segments(self, osm_route: Route) -> List[Segment]:
+        """Prepare list of segments based on route."""
+        result = []
+        for u, v in zip(osm_route[:-1], osm_route[1:]):
+            # if there are parallel edges between two nodes, select the one with the
```
Comment (Kobzol) [inline ruth/data/map.py:109]: There should be no parallel edges, because `self.simple_network` is already a `DiGraph`, not a `MultiDiGraph`, right? Why was this added?

File: ruth/vehicle.py [hunk 4]
```diff
@@ -151,5 +156,9 @@ def set_position(self, sp: SegmentPosition):
     def is_active(self, within_offset, freq):
         return self.active and within_offset == round_timedelta(self.time_offset, freq)
 
+    def is_at_the_end_of_segment(self):
```
[the remark this answers is not in the document]
Comment (Kobzol) [reply ruth/vehicle.py:159]: This is still quite hacky. Can't we just ask the map for a single segment, given `from`/`to` IDs from `self.osm_route[self.start_index]`?

File: ruth/data/map.py (hunk 3 above)
Comment (frackledust) [reply ruth/data/map.py:109]: Ah yes. My bad. The comment is not necessary here. The functionality was taken from the get_route_edge_attributes since that function is getting deprecated (and the suggested alternative is not useful for our usecase),

File: ruth/data/map.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (frackledust) [reply ruth/data/map.py:94]: Yeah, cpp probably computes the edge weights by itself. One difference is for example that it ignores edges with speed 0 (which turned out to be a bit issue for the version with updated map, but we are working on it).

Based on some py-spy measurements we took the update_current_speed function takes about 2 seconds in total, so it's not that bad.

[the remark this answers is not in the document]
Comment (Kobzol) [reply ruth/data/map.py:97]: 2 seconds for how many edges updated?

[the remark this answers is not in the document]
Comment (frackledust) [reply ruth/data/map.py:97]: Cannot say rn, but 2 seconds for cca 45 minutes of simulation running, if I understand the py-spy record correctly and the map being updated every next step of the simulation. The most time consuming in the simulation rn is computing the alternatives (then saving the map and the simulation I think).

[the remark this answers is not in the document]
Comment (Kobzol) [reply ruth/data/map.py:97]: Sorry, I missed the `in total` part. Then it should be fine.

File: ruth/globalview.py [hunk 5]
```diff
@@ -53,8 +53,13 @@ def level_of_service_for_car(self, datetime, segment, vehicle: Vehicle, toleranc
 
         n_vehicles = self.number_of_vehicles_in_time_at_segment(datetime, segment.id, tolerance, vehicle)
 
+        ending_length = 100
+        rest_segment_length = segment.length - vehicle.start_distance_offset
```
[the remark this answers is not in the document]
Comment (Michailidu) [reply ruth/globalview.py]: The number of vehicles per mile calculated below is higher the closer the vehicle is to the end of the segment (if there are vehicles in front of it). This causes lags at the ends of the segments and also on short segments. Setting different calculation for the last 100 (now 200) meters is a temporary solution. We will probably use multiple scales for the LoS as a long-term solution instead of this one. Should we add this functionality to this PR or is it better to change it back and keep the old behaviour for now?

Also, `number_of_vehicles_in_time_at_segment` should return only the number of vehicles that are in the rest of the segment in this case.

[the remark this answers is not in the document]
Comment (Kobzol) [reply ruth/globalview.py]: > Also, number_of_vehicles_in_time_at_segment should return only the number of vehicles that are in the rest of the segment in this case.

Ok, this was the missing piece. This is highly non-obvious from the code. When `vehicle_id` and `vehicle_offset_m` are used, the name of the function `number_of_vehicles_in_time_at_segment` isn't really accurate. I would just create another function for this, because it's quite confusing.

File: ruth/simulator/singlenode.py [hunk 6]
```diff
@@ -52,28 +52,30 @@ def simulate(
         alternatives_provider.load_map(self.sim.routing_map)
 
         step = self.sim.number_of_steps
+        if self.sim.setting.speeds_path is not None:
```
[the remark this answers is not in the document]
Comment (Kobzol) [reply ruth/simulator/singlenode.py:57]: :+1: Great.

File: ruth/simulator/segment.py [hunk 7]
```diff
@@ -15,6 +15,9 @@ class SegmentPosition:
     # Position within the segment (in meters)
     position: LengthMeters
 
+    def __eq__(self, other):
```
Comment (Kobzol) [inline ruth/simulator/segment.py:18]: This is implemented by default for dataclasses, you don't need to implement it manually.

**Review by Kobzol — APPROVED** [review]
There's one additional nit, and it would be good to remove/rename the WIP commit messages, but other than that LGTM.
